enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+// IPAddressStr is used for validation of an IP address.
+type IPAddressStr string
+
+// DNSNameResolverStatusItem describes the details of a resolved DNS name.
+type DNSNameResolverStatusItem struct {
+	// conditions provide information about the state of the DNS name.
+	// Known .status.conditions.type is: "Degraded"
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+	// dnsName is the resolved DNS name matching the name field of DNSNameResolverSpec. This field can
+	// store both regular and wildcard DNS names which match the spec.name field. When the spec.name
+	// field contains a regular DNS name, this field will store the same regular DNS name after it is
+	// successfully resolved. When the spec.name field contains a wildcard DNS name, this field will
+	// store the regular DNS names which match the wildcard DNS name and have been successfully resolved.
+	// If the wildcard DNS name can also be successfully resolved, then this field will store the wildcard
+	// DNS name as well.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	// +kubebuilder:validation:MaxLength=254
+	DNSName string `json:"dnsName"`
+	// resolvedAddresses gives the list of associated IP addresses and their corresponding TTLs and last
+	// lookup times for the dnsName.
+	// +kubebuilder:validation:Required
+	// +listType=map
+	// +listMapKey=ip
+	ResolvedAddresses []DNSNameResolverInfo `json:"resolvedAddresses"`
+	// resolutionFailures keeps the count of how many consecutive times the DNS resolution failed
+	// for the dnsName. If the DNS resolution succeeds then the field will be set to zero. Upon
+	// every failure, the value of the field will be incremented by one. The details about the DNS
+	// name will be removed, if the value of resolutionFailures reaches 5 and the TTL of all the
+	// associated IP addresses have expired.
+	ResolutionFailures int32 `json:"resolutionFailures,omitempty"`
+}
+
+type DNSNameResolverInfo struct {
+	// ip is an IP address associated with the dnsName. The validity of the IP address expires after
+	// lastLookupTime + ttlSeconds. To refresh the information a DNS lookup will be performed on the
+	// expiration of the IP address's validity. If the information is not refreshed then it will be
+	// removed after a grace period of 1 second after the expiration of the IP address's validity.
+	// +kubebuilder:validation:Required
+	IP IPAddressStr `json:"ip"`
+	// ttlSeconds is the time-to-live value of the IP address.
+	// +kubebuilder:validation:Required
+	TTLSeconds int32 `json:"ttlSeconds"`
+	// lastLookupTime is the timestamp when the last DNS lookup was completed.
+	// +kubebuilder:validation:Required
+	LastLookupTime *metav1.Time `json:"lastLookupTime"`
+}
+````
+
+### Workflow Description
```

Comment:
I'd like to see the the steps from the point of view of the DNS lookup - which steps are taken when there is an Egress FW, when there isn't, when there is a wildcard, when there isn't, when a DNS name is in multiple Egress FW, and when there is an address not even in Egress FW, what's the flow?

Reply:
The DNS lookup details is added in the workflow of the plugin: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R333-R500